src/raftstore/store/snap/io.rs

```diff
@@ -0,0 +1,246 @@
+    let mut stats = BuildStatistics::default();
+    box_try!(snap.scan_cf(cf, start_key, end_key, false, |key, value| {
+        stats.key_count += 1;
+        stats.total_size += key.len() + value.len();
+        box_try!(file.encode_compact_bytes(key));
+        box_try!(file.encode_compact_bytes(value));
+        Ok(true)
+    }));
+    if stats.key_count > 0 {
+        // use an empty byte array to indicate that cf reaches an end.
+        box_try!(file.encode_compact_bytes(b""));
+        box_try!(file.sync_all());
+    } else {
+        drop(file);
+        box_try!(fs::remove_file(path));
+    }
+    Ok(stats)
+}
+
+/// Build a snapshot file for the given column family in sst format.
+/// If there are no key-value pairs fetched, no files will be created at `path`,
+/// otherwise the file will be created and synchronized.
+pub fn build_sst_cf_file(
+    path: &str,
+    snap: &DbSnapshot,
+    cf: &str,
+    start_key: &[u8],
+    end_key: &[u8],
+    io_limiter: Option<&IOLimiter>,
+) -> Result<BuildStatistics, Error> {
+    let mut sst_writer = create_sst_file_writer(snap, cf, path)?;
+    let mut stats = BuildStatistics::default();
+    let mut base = if io_limiter.is_some() { 0 } else { usize::MAX };
+    box_try!(snap.scan_cf(cf, start_key, end_key, false, |key, value| {
+        if let Some(ref io_limiter) = io_limiter {
+            if stats.total_size >= base {
+                let next_request = io_limiter.get_max_bytes_per_time();
+                io_limiter.request(next_request);
+                base += next_request as usize;
+            }
+        }
+        stats.key_count += 1;
+        stats.total_size += key.len() + value.len();
+        if let Err(e) = sst_writer.put(key, value) {
+            let io_error = io::Error::new(io::ErrorKind::Other, e);
+            return Err(io_error.into());
+        }
+        Ok(true)
+    }));
+    if stats.key_count > 0 {
+        box_try!(sst_writer.finish());
+        drop(sst_writer);
+        box_try!(File::open(path).and_then(|f| f.sync_all()));
+    } else {
+        drop(sst_writer);
+        box_try!(fs::remove_file(path));
+    }
+    Ok(stats)
+}
+
+/// Apply the given snapshot file into a column family.
+pub fn apply_plain_cf_file(
+    path: &str,
+    stale_detector: &impl StaleDetector,
+    db: &DB,
+    cf: &str,
+    batch_size: usize,
+) -> Result<(), Error> {
+    let mut decoder = BufReader::new(box_try!(File::open(path)));
+    let cf_handle = box_try!(get_cf_handle(&db, cf));
+    let wb = WriteBatch::new();
+    loop {
+        if stale_detector.is_stale() {
+            return Err(Error::Abort);
+        }
+        let key = box_try!(decoder.decode_compact_bytes());
+        if key.is_empty() {
+            box_try!(db.write(&wb));
+            return Ok(());
+        }
+        let value = box_try!(decoder.decode_compact_bytes());
+        box_try!(wb.put_cf(cf_handle, &key, &value));
```

Comment:
Plese check key in region first.

Reply:
Seems the key in region check do not exist even in the master branch.